lib/autoproj/daemon/buildconf_manager.rb

```diff
@@ -0,0 +1,250 @@
+                created
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Array<PackageRepository>]
+            def packages_affected_by_pull_request(pull_request)
+                packages.select do |pkg|
+                    pkg.name == pull_request.base_name &&
+                        pkg.owner == pull_request.base_owner &&
+                        pkg.branch == pull_request.base_branch
+                end
+            end
+
+            # @param [Github::PullRequest] pull_request
+            # @return [Hash]
+            def overrides_for_pull_request(pull_request)
+                # TODO: Extract overrides from PR body.
+                #    This is not as easy as one could have thought
+                #    because the depended upon PR may have dependencies
+                #    of its own. We have to do it recursively. Detecting
+                #    changes in this chain of overrides is even more
+                #    challenging. Or, one could add the entire dependency
+                #    tree in every PR body.
+                packages_affected_by_pull_request(pull_request).map do |pkg|
+                    key = if pkg.package_set?
+                              "pkg_set:#{pkg.vcs[:repository_id]}"
+                          else
+                              pkg.package
+                          end
+                    {
+                        key => {
+                            'github' => "#{pull_request.head_owner}/"\
+                                        "#{pull_request.head_name}",
+                            'remote_branch' => pull_request.head_branch
+                        }
+                    }
+                end
+            end
+
+            # @param [Github::Branch] branch
+            # @return [Github::PullRequest, nil]
+            def pull_request_by_branch(branch)
+                pull_requests.find do |pr|
+                    branch_name_by_pull_request(pr) == branch.branch_name
+                end
+            end
+
+            # @param [Github::Branch] branch
+            # @return [nil]
+            def trigger_build(branch)
+                Autoproj.message "Triggering build for #{branch.branch_name} (new)"
+                nil
+            end
+
+            # @param [Array<Github::Branch>] branches
+            # @return [nil]
+            def trigger_build_if_branch_changed(branches)
+                branches.each do |branch|
+                    pr = pull_request_by_branch(branch)
+                    overrides = overrides_for_pull_request(pr)
+                    next unless cache.changed?(pr, overrides)
+
+                    cache.add(pr, overrides)
+                    Autoproj.message "Triggering build for #{branch.branch_name} (update)"
+                end
+                nil
+            end
+
+            # @return [nil]
+            def synchronize_branches
+                cache.clear
```

Comment:
An overall programming style that I prefer is to make most of the class methods "functional", that is they wouldn't modify/rely on the value of the internal attributes of the class, and have `synchronize_branches` pass and store the values around, for instance:

~~~
cache.clear
@pull_requests = update_pull_requests
@branches = update_branches
delete_branches(find_stale_branches)
...
~~~

This is a lot easier to test, avoiding the need to expose so much of the internal class state.

Reply:
I don't really have a preference on this one. Made sense to me when I wrote as is but I can change if you believe that would be an improvement. The methods would have to be renamed to `fetch_` rather than `update_` though

Reply:
Another reason for writing the way I did was to reduce the complexity (and make rubocop happy). These warnings (about complexity) are kinda boring and take some getting used to but I really think they have a huge impact on code quality and readability....

Reply:
I don't think that you will have more complexity warnings with the functional style, and it IMO makes it a lot easier to reason about the data flow since most methods won't change the object's state. I think it is a good general design guideline (favoring less side-effects). I won't push too hard on this, so follow what you believe is best.